bpf: misc ICMPv6 ND fixes (srcip, nollsrcopt...)

msune opened the pull request:
(Stubbed / dependency of PR #34983)

This patchset:

```
ebead7f9c0 bpf/{lib|tests}: fix ICMPv6 NS w/o LLSRC handling
ce604f1096 bpf: add ICMPv6 ND mcast coverage
bbdd2ac0e8 bpf/{lib|test}: send ICMP NA from target SRC_IP
8f433835a2 bpf/lib/icmpv6: remove ifdef IPV6
```

See commit logs for details.

```release-note
Improve ICMPv6 Neighbor Discovery packet handling
```

msune committed:
bpf/{lib|test}: send ICMP NA from target SRC_IP

Prior to this commit, Neighbour Solicitation replies
(Neighbour Advertisments) were sent from ROUTER_IP.

RFC4861, section 4.4 explicitly mentions that the Source Address
must be:

> An address assigned to the interface from which the advertisement
> is sent.

In addition, the current BPF ARP implementation returns the ARP
header with the target_ip as source IP.

This commit changes the current implementation of icmp6_send_reply()
to use the target_ip as source IP of ICMP NA messages.

Signed-off-by: Marc Suñé <marc.sune@isovalent.com>
Signed-off-by: Timo Beckers <timo@isovalent.com>

msune committed:
bpf: add ICMPv6 ND mcast coverage

This commit refactors and adds two more tests to ipv6_ndp_from_netdev_test.c
covering multicast (broadcast) IPv6 neighbour solicitation, both for node IPs
(to stack) and for endpoints (BPF).

Signed-off-by: Marc Suñé <marc.sune@isovalent.com>
Signed-off-by: Timo Beckers <timo@isovalent.com>

msune committed:
bpf/{lib|tests}: fix ICMPv6 NS w/o LLSRC handling

Prior to this commit, the ICMPv6 Neighbour Solicitation BPF responder (Pods)
implementation assumed that the Link Layer Source Address option was
**always** present in the packet.

According to spec, this option is - actually - optional for NS messages,
and SHALL NOT be used in particular cases:

> According to RFC4861, sections 4.3 and 7.2.2 unicast neighbour
> solicitations (reachability check) SHOULD but are NOT REQUIRED to
> include the SRC_LL_ADDR option in the NS message.
>
> Likewise, neighbour solicitations during Duplicate Address Detection
> (DAD, RFC4862), SRC_LL_ADDR option must not present.

This resulted in NS messages being dropped when attempting to access LLSRC
option (out of bounds check).

This commit:

* Fixes handling of ICMPv6 NS packets by allocating via ctx_change_tail()
  enough space for LL_SRC option when not originally present.
* Adds test coverage for all combinations Targeted/Bcast, Node/Pod and
  w/ and w/o LL_SRC option.

Signed-off-by: Marc Suñé <marc.sune@isovalent.com>
Signed-off-by: Timo Beckers <timo@isovalent.com>

gentoo-root commented on bpf/tests/ipv6_ndp_from_netdev_test.c:
```diff
@@ -59,16 +65,17 @@ int ipv6_from_netdev_ns_for_pod_pktgen(struct __ctx_buff *ctx)
 {
 	endpoint_v6_add_entry((union v6addr *)v6_pod_three, 0, 0, 0, 0,
 			      (__u8 *)mac_three, (__u8 *)mac_two);
+	printk("-- Test start : 011_ipv6_from_netdev_ns_for_pod");
```
We don't seem to print these messages in other tests. Doesn't the test framework print some markers at the beginning and in the end of each test by default?

msune replied:
Humm.. I am not aware of it, but happy if it's there already!

The existing test was printing _some_ traces (e.g. https://github.com/cilium/cilium/blob/main/bpf/tests/ipv6_ndp_from_netdev_test.c#L91), so I thought adding a bit more context would be useful to see which of the permutations was failing.

gentoo-root replied:
> I am not aware of it, but happy if it's there already!

I'm not 100% sure of that, but I think I saw such markers (unless I mixed it up with some other testset :thinking:). Either way, I'd say it's more preferable to add these prints to the test framework if they are missing, rather than to each individual test.

> The existing test was printing some traces

That seems to be the only two occurrences (lines 91 and 212) of printk in the entire testset, both added in the same commit. If I reviewed that, I would suggest to introduce an `assert_eq` function that would print the mismatching values in case of an error, rather than adding extra prints at each `assert(... == ...)` and then digging through logs to find the relevant one.

msune replied:
> That seems to be the only two occurrences (lines 91 and 212) of printk in the entire testset, both added in the same commit

That's fair; all removed, including previous traces.

gentoo-root commented on bpf/lib/icmp6.h:
```diff
@@ -87,25 +88,68 @@ int icmp6_send_reply(struct __ctx_buff *ctx, int nh_off, union v6addr new_sip)
+							+ sizeof(union v6addr));
+	if ((void *)(opt + 1) > data_end)
+		return DROP_INVALID;
+	*opt = 0x0ULL;
```
Just to make sure I understand the idea correctly: we don't actually care about the SRC_LL_ADDR option that came with the incoming packet, but we reuse that space to add TARGET_LL_ADDR to the outgoing packet, so you are reserving that space if SRC_LL_ADDR was missing — right?

msune replied:
The original packet doesn't have SRC_LL_ADDR, that's why it's making it to this function.

Per RFC, [NS message](https://datatracker.ietf.org/doc/html/rfc4861#section-4.3) must have the target IPv6 address, always, but  SRC_LL_ADDR (which is to pre-fill the ARP table when processing NS) is optional, and in some cases - as discussed in the comment - should NOT be there (e.g. Duplicate Address Detection or DAD) .

Here I just reserve space and 0 it. 0ing it makes the code path down in the `send_reply()` easier, as it always reads the _previous_ option value to compute the csum diff with the new option value.

gentoo-root replied:
Yes, I understand these points, thanks. Just wanted to make sure we aren't going to accidentally treat this dummy 0 as a real SRC_LL_ADDR.

msune replied:
Nope, the ND target SRC_LL_ADR option is always overwritten here with the right value (hence the reason pkts without it were discarded):

https://github.com/cilium/cilium/pull/36024/files#diff-05a04bf67c393d1c78d8b25bf5e858dc62bed8594a7a4b09ea26983c3b0d5cb0R193-R204

gentoo-root commented on bpf/tests/ipv6_ndp_from_netdev_test.c:
```diff
@@ -181,38 +229,44 @@ int ipv6_from_netdev_ns_for_pod_pktgen_mcast(struct __ctx_buff *ctx)
-int ipv6_from_netdev_ns_for_pod_setup_mcast(struct __ctx_buff *ctx)
+static __always_inline
+int __ipv6_from_netdev_ns_for_pod_setup_mcast(struct __ctx_buff *ctx,
+								bool llsrc_opt)
```
Some continued lines are misaligned here and there, but probably checkpatch will point you at all occurrences.

msune replied:
checkpatch doesn't warn or error, but has a `CHECK` on it. I was trying this:

```
int __ipv6_from_netdev_ns_for_pod_setup_mcast(struct __ctx_buff *ctx,
						bool llsrc_opt)
```

But wasn't quite happy; what would be the correct alignment in this example?

gentoo-root replied:
It should be aligned to `(` with the maximum number of tabs + the remainder padded with spaces (each tab is 8-character wide). 5 tabs + 6 spaces in this example:

```
int __ipv6_from_netdev_ns_for_pod_setup_mcast(struct __ctx_buff *ctx,
					      bool llsrc_opt)
```

(I'm not a fan of fixed-width tabs, but) that's the style we use here, because it's what the [kernel uses](https://www.kernel.org/doc/html/latest/process/coding-style.html):

```
Tabs are 8 characters, and thus indentations are also 8 characters.
A very commonly used style is to align descendants to a function open parenthesis.
These same rules are applied to function headers with a long argument list.
```

(Though, it doesn't seem to mention using spaces after tabs to pad the remainder mod 8, you can check any long function header in Linux or Cilium to check that it's indeed used this way.)

msune replied:
I think I got confused by the output of `make checkpatch` for uncommited changes (I've used `checkpatch.pl` long time ago to contrib to DPDK), and what the command was telling me is that a previous commit (as I didn't commit&squash) had still those indents.

Should be fixed now, as the checkpatch workflow passed.

gentoo-root reviewed (approved):
Looks good to me in general; I haven't analyzed the CI failures (yet?).

ti-mo commented on bpf/lib/icmp6.h:
```diff
@@ -1,7 +1,7 @@
 /* Copyright Authors of Cilium */
 
-#if !defined(__LIB_ICMP6__) && defined(ENABLE_IPV6)
+#ifndef __LIB_ICMP6__
```
Nit: change this `#ifndef` and subsequent `#define` to `#pragma once` for simplicity.

msune replied:
Humm.. there is a mix of `#pragma once` and `#ifndef` as guards in `lib/`, and in general across the codebase. I am personally biased towards `#ifndef` (PTS?). Is there a consensus on this?

msune replied:
@ti-mo :arrow_up: ?

ti-mo reviewed (commented):
> I can't see anything "problematic" in `icmpv6.h`, so I suspect it has more to do with the tests themselves. I would appreciate some guidance on this, cc @ti-mo.

There's indeed nothing problematic in `linux/icmpv6.h`. `fib_tests.c` has had some data in .bss since forever, redir_recorder and redir_neigh_recorder are global vars:

```
~ llvm-objdump -SD bpf/tests/fib_tests.o
[...]
Disassembly of section .bss:

0000000000000000 <redir_recorder>:
...

0000000000000018 <redir_neigh_recorder>:
...
```

The inliner happily accepted it since bpf tests all override configurables like `THIS_INTERFACE_MAC` using compile-time macros, meaning there are no references to `.rodata.config`. This makes the following statement return early: https://github.com/cilium/cilium/blob/11da06e24d3a2d4e7c9e37a8b6f7f984aa662752/pkg/bpf/collection.go#L437-L441

.. so the .bss reference doesn't trigger the error.

With the guard removed, `lib/icmp6.h` is compiled in unconditionally, creating a reference to `.rodata.config`, triggering the inliner. I'm removing this restriction in https://github.com/cilium/cilium/pull/35988 if you can wait.

Keep in mind the removal of the ifdef cannot be backported for this reason, so best to drop the change or send it as a separate PR. It will need to happen at some point if we ever want to go clang-free at runtime anyway.

msune committed:
bpf: refactor ICMPv6 ND tests

Remove redundancy in test code for ICMPv6, partially existent and
partially introduced in previous commits. Add some more assertions
e.g. ICMPv6 type/length.

Signed-off-by: Marc Suñé <marc.sune@isovalent.com>
Signed-off-by: Timo Beckers <timo@isovalent.com>

dylandreimerink reviewed (approved):
I believe this all makes sense, can't find anything wrong :rocket:

rgo3 reviewed (commented):
Looks like I don't own any of the files in this PR, so removing myself as reviewer. As a side note, this PR will probably need a rebase after https://github.com/cilium/cilium/pull/38244 goes in (would also solve the checkpatch complaint I think). Or vice versa of course.
